feat: add & remove addresses from accounts

juan518munoz opened the pull request:
Closes #1271

SantiagoPittella commented on bin/miden-cli/src/commands/addresses.rs:
Lets rename it to singular

juan518munoz commented on bin/miden-cli/src/commands/addresses.rs:
```diff
@@ -0,0 +1,135 @@
+    info!(
+        "removing address - Account Id {} - Execution mode: {}",
+        account_id, execution_mode
+    );
```
Addressed in [f769ecf](https://github.com/0xMiden/miden-client/pull/1367/commits/f769ecf9f9a78d92cb3b1daa04a29c5d3cc3a1fc)

juan518munoz commented on bin/miden-cli/src/commands/addresses.rs:
```diff
@@ -0,0 +1,135 @@
+        NoteExecutionMode::Local => "Local",
+        NoteExecutionMode::Network => "Network",
+    };
+    info!(
```
Addressed in [f769ecf](https://github.com/0xMiden/miden-client/pull/1367/commits/f769ecf9f9a78d92cb3b1daa04a29c5d3cc3a1fc)

fkrause98 commented on bin/miden-cli/src/commands/addresses.rs:
```diff
@@ -0,0 +1,131 @@
+            },
+        }
+        Ok(())
+    }
```
Handling multiple boolean flags can get messy quick, why don't we do something like this?

```suggestion
    #[derive(Debug, Subcommand, Clone)]
pub enum AddressSubCommand {
    /// List all addresses an account can be referenced by
    List { account_id: String },
    /// Add a new address,
    Add {
        /// Interface number for add/remove operations
        interface: String,
        /// Account to Add
        account_id: String,
    },
    /// Remove the given address.
    Remove {
        /// Interface number for add/remove operations
        interface: String,
        /// Account to Remove
        account_id: String,
    },
}

#[derive(Debug, Parser, Clone)]
#[command(about = "Manage account addresses")]
pub struct AddressesCmd {
    #[clap(subcommand)]
    command: AddressSubCommand,
}

impl AddressesCmd {
    pub async fn execute<AUTH>(&self, client: Client<AUTH>) -> Result<(), CliError> {
        match &self.command {
            AddressSubCommand::List { account_id } => {
                list_addresses(client, account_id.clone()).await?
            },
            AddressSubCommand::Add { interface, account_id } => {
                remove_address(client, account_id.clone(), interface.clone()).await?;
            },
            AddressSubCommand::Remove { interface, account_id } => {
                add_address(client, account_id.clone(), interface.clone()).await?;
            },
        }
        Ok(())
    }
}
```

fkrause98 reviewed (commented):
Let's add some tests for the new TS code.

juan518munoz commented on bin/miden-cli/src/commands/addresses.rs:
```diff
@@ -0,0 +1,130 @@
+pub enum AddressSubCommand {
+    /// List all addresses an account can be referenced by
+    List { account_id: String },
+    /// Add a new address,
```
Done!

mmagician commented on bin/miden-cli/src/commands/addresses.rs:
```diff
@@ -0,0 +1,130 @@
+    };
+
+    println!("Addresses for AccountId {account_id}:");
+    let mut table = create_dynamic_table(&["Address", "Interface"]);
```
in which case the table would include `AccountId` too

juan518munoz replied:
May I suggest that we list a separate table for each `Account`? Like this:

<img width="1229" height="469" alt="image" src="https://github.com/user-attachments/assets/0e9e5a1c-f9ff-410c-b8c9-fc533e31b00c" />

juan518munoz replied:
Went with proposed solution in [e1f5cbc](https://github.com/0xMiden/miden-client/pull/1367/commits/e1f5cbc98c7342e515dc7a49aec4d3dcc27fa19a)

mmagician commented on bin/miden-cli/src/commands/addresses.rs:
```diff
@@ -0,0 +1,130 @@
+) -> Result<Address, CliError> {
+    let interface = match interface {
+        "unspecified" => AddressInterface::Unspecified,
+        "basic_wallet" => AddressInterface::BasicWallet,
```
this should be consistent with the strings returned by the list command

juan518munoz replied:
Addressed in [e1f5cbc](https://github.com/0xMiden/miden-client/pull/1367/commits/e1f5cbc98c7342e515dc7a49aec4d3dcc27fa19a)

juan518munoz commented on bin/miden-cli/src/commands/addresses.rs:
```diff
@@ -0,0 +1,130 @@
+    println!("Addresses for AccountId {account_id}:");
+    let mut table = create_dynamic_table(&["Address", "Interface"]);
+    for address in addresses {
+        let address_hex = hex::encode(address.to_bytes());
```
Replaced for bech32 in [e1f5cbc](https://github.com/0xMiden/miden-client/pull/1367/commits/e1f5cbc98c7342e515dc7a49aec4d3dcc27fa19a)

igamigo commented on crates/rust-client/src/account/mod.rs:
```diff
@@ -221,6 +222,50 @@ impl<AUTH> Client<AUTH> {
+        &mut self,
+        address: Address,
+        account_id: AccountId,
+    ) -> Result<(), ClientError> {
```
Should we check that if `address` encodes an `AccountId`, that it is the same as `account_id`? Otherwise (more of a question for @mmagician), what would be the benefit of associating an address of a different account ID to an arbitrary one? I think this would also apply in terms of interface, wouldn't we know from knowing the account code which interfaces are potentially associated to the account?

juan518munoz replied:
`Address` has no accesor to its inner `AccountId` AFAIK. Should we open a PR in `miden-base` addressing this?

mmagician replied:
I am also ok with leaving this as-is for now, since it will change with the upcoming `Address`-related changes

igamigo commented on bin/miden-cli/src/commands/addresses.rs:
```diff
@@ -0,0 +1,156 @@
+        /// Account to add
+        account_id: String,
+        /// Interface number for add/remove operations
+        interface: String,
```
Is there a way we can show what interfaces are supported here? Otherwise, it might be hard for users to know what they could pass to this. Maybe we should create our own `CliAddressInterface` and add transformations for `AddressInterface` so that the values appear on the help descriptions

igamigo commented:
I also wonder if for the CLI command (and maybe this is a question for @mmagician as well), if we should make it so that the "add address" command takes only `AccountId` and `Address` (encoded as string). And then separately we could have a util command that outputs addresses based on fields (and maybe also parses and tells you what an address is encoding)? If this makes sense we don't have to do it here, but it feels like a slightly better UX for these commands.

juan518munoz commented on bin/miden-cli/src/commands/addresses.rs:
```diff
@@ -0,0 +1,156 @@
+#[derive(Debug, Subcommand, Clone)]
+pub enum AddressSubCommand {
+    /// List all addresses an account can be referenced by
+    List { account_id: Option<String> },
```
Done in [9039a2f](https://github.com/0xMiden/miden-client/pull/1367/commits/9039a2fac8a60001f449c2cdc29ccb3e4caeb1f7)

mmagician commented:
> I also wonder if for the CLI command (and maybe this is a question for @mmagician as well), if we should make it so that the "add address" command takes only AccountId and Address (encoded as string). And then separately we could have a util command that outputs addresses based on fields

That's a good idea. I would, however, leave that for a follow-up PR. Especially for now it is ok since we don't have _that many fields yet_.

mmagician commented:
> EDIT: @mmagician the error should now be more reflective of the error:

@juan518munoz thanks for changing the error to include the info about the tag, that's indeed helpful.

I think what is more confusing is that the error mentions an `Address` which is not displayed in the table. (btw, is the table correct? currently it says:
```
Addresses for AccountId mlcl1qqkn9dvfdkhg7yr0jpt2mraj5ecqqjtgq05:

// within the table:
Address: mlcl1qqkn9dvfdkhg7yr0jpt2mraj5ecqqjtgq05
```

whereas the representation of `Address` and `AccountId` should be different.

juan518munoz commented on crates/idxdb-store/src/js/accounts.js:
Addressed in [392a351](https://github.com/0xMiden/miden-client/pull/1367/commits/392a351f8cceda00d9a7c6c609073808a9b684c3)

juan518munoz commented:
> > I also wonder if for the CLI command (and maybe this is a question for @mmagician as well), if we should make it so that the "add address" command takes only AccountId and Address (encoded as string). And then separately we could have a util command that outputs addresses based on fields
> 
> That's a good idea. I would, however, leave that for a follow-up PR. Especially for now it is ok since we don't have _that many fields yet_.

Opened https://github.com/0xMiden/miden-client/issues/1432

mmagician commented:
@juan518munoz I think the only leftover comment I have is [this one](https://github.com/0xMiden/miden-client/pull/1367#issuecomment-3444484199)

juan518munoz commented:
@mmagician thanks for the heads up, I forgot about them.

> @juan518munoz I think the only leftover comment I have is [this one](https://github.com/0xMiden/miden-client/pull/1367#issuecomment-3444484199)

> (btw, is the table correct? currently it says:
Addresses for AccountId mlcl1qqkn9dvfdkhg7yr0jpt2mraj5ecqqjtgq05:
// within the table:
Address: mlcl1qqkn9dvfdkhg7yr0jpt2mraj5ecqqjtgq05
whereas the representation of Address and AccountId should be different.

That's because the way we parse the input `AccountId` is with the function [`parse_account_id`](https://github.com/0xMiden/miden-client/blob/900dc3f6b97ba5d2f8bd047c690b816d4c6c0e8a/bin/miden-cli/src/utils.rs#L54), which takes multiple inputs as valid:
```rs
/// `account_id` can fall into three categories:
///
/// - It's a hex prefix of an account ID of an account tracked by the client.
/// - It's a full hex account ID.
/// - It's a full bech32 account ID.
```

Nevertheless, I have updated the function so even if it can take either representation as valid, when listing the tables it displays in the `hex` form:
```bash
➜  miden-client git:(jmunoz-dynamic-address-for-accounts) ✗ miden-client address list mlcl1qqkn9dvfdkhg7yr0jpt2mraj5ecqqjtgq05
Addresses for AccountId 0x2d32b5896dae8f106f9056ad8fb2a6:
┌──────────────────────────────────────────────────────────────────────────────────────┬─────────────────────────────────────────────────────────┐
│ Address                                                                              ┆ Interface                                               │
╞══════════════════════════════════════════════════════════════════════════════════════╪═════════════════════════════════════════════════════════╡
│ mlcl1qqkn9dvfdkhg7yr0jpt2mraj5ecqqjtgq05                                             ┆ Unspecified                                             │
└──────────────────────────────────────────────────────────────────────────────────────┴─────────────────────────────────────────────────────────┘
```

> I think what is more confusing is that the error mentions an Address which is not displayed in the table. 

I've changed the error message a bit further, so it's clearer if the failure is due to a note tag or address already being tracked:
```bash
➜  miden-client git:(jmunoz-dynamic-address-for-accounts) ✗ miden-client address list 0x2d32b5896dae8f106f9056ad8fb2a6
Addresses for AccountId 0x2d32b5896dae8f106f9056ad8fb2a6:
┌──────────────────────────────────────────────────────────────────────────────────────┬─────────────────────────────────────────────────────────┐
│ Address                                                                              ┆ Interface                                               │
╞══════════════════════════════════════════════════════════════════════════════════════╪═════════════════════════════════════════════════════════╡
│ mlcl1qqkn9dvfdkhg7yr0jpt2mraj5ecqqjtgq05                                             ┆ Unspecified                                             │
└──────────────────────────────────────────────────────────────────────────────────────┴─────────────────────────────────────────────────────────┘
➜  miden-client git:(jmunoz-dynamic-address-for-accounts) ✗ miden-client address add 0x2d32b5896dae8f106f9056ad8fb2a6 Unspecified
Error: cli::client_error

  × client error
  ╰─▶ address mlcl1qqkn9dvfdkhg7yr0jpt2mraj5ecqqjtgq05 is already being tracked

➜  miden-client git:(jmunoz-dynamic-address-for-accounts) ✗ miden-client address add 0x2d32b5896dae8f106f9056ad8fb2a6 BasicWallet
Error: cli::client_error

  × client error
  ╰─▶ address mlcl1qqkn9dvfdkhg7yr0jpt2mraj5ecqzpwfrue available, but its derived note tag 3410755584 is already being tracked
```

Addressed in [d54458e](https://github.com/0xMiden/miden-client/pull/1367/commits/d54458e65eff16cced465051bc96819ead88ee3b)

mmagician reviewed (approved):
That's a great change to the error message displayed, thanks!
